src/microbe_stage/MembraneWaterRipple.cs

```diff
@@ -0,0 +1,666 @@
+        {
+            waterMaterial = material;
+            planeMesh = mesh;
+
+            // Initialize the position history arrays
+            godotPastPositions.Clear();
+            for (int i = 0; i < PositionHistorySize; ++i)
+            {
+                godotPastPositions.Add(Vector2.Zero);
+                pastPositions[i] = Vector2.Zero;
+                positionHistory[i] = Vector3.Zero;
+            }
+
+            // Initialize state
+            isEffectEnabled = true;
+            currentAlpha = 0.0f;
+            targetAlpha = 0.0f;
+            stillnessValue = 1.0f;
+            targetStillness = 1.0f;
+            isForming = false;
+            formingTime = 0.0f;
+            stillnessFactor = 0.0f;
+            stillnessTimer = 0.0f;
+            timeWithoutMovement = 0.0f;
+            currentPositionIndex = 0;
+            isPositionHistoryFull = false;
+
+            // Initialize shader parameters
+            waterMaterial.SetShaderParameter(waterColorParam, new Color(0, 0, 0, 0.02f));
+            waterMaterial.SetShaderParameter(rippleStrengthParam, RippleStrength);
+            waterMaterial.SetShaderParameter(timeOffsetParam, 0.0f);
+            waterMaterial.SetShaderParameter(movementSpeedParam, 0.0f);
+            waterMaterial.SetShaderParameter(movementDirectionParam, Vector2.Zero);
+            waterMaterial.SetShaderParameter(phaseParam, 0.2f);
+            waterMaterial.SetShaderParameter(attenuationParam, 0.9998f);
+            waterMaterial.SetShaderParameter(stillnessFactorParam, 0.0f);
+            waterMaterial.SetShaderParameter(pastPositionsParam, godotPastPositions);
+            waterMaterial.SetShaderParameter(pastPositionsCountParam, 0);
+            waterMaterial.SetShaderParameter(membraneRadiusParam, 5.0f);
+            currentCamera = GetViewport().GetCamera3D();
+            waterPlane.Visible = false;
+
+            // If we got here everything is properly initialized :)
+            isInitialized = true;
+        }
+    }
+
+    public override void _Process(double delta)
+    {
+        // Skip if not initialized or missing required components
+        if (!isInitialized || !EnableEffect || positionTarget == null)
+            return;
+
+        // Clamp delta to further prevent jittering
+        float clampedDelta = MathF.Min((float)delta, MAX_DELTA_TIME);
+        currentCamera = GetViewport().GetCamera3D();
+
+        // Updates camera and visibility
+        visibilityCheckTimer += clampedDelta;
+        if (visibilityCheckTimer >= VisibilityCheckInterval)
+        {
+            visibilityCheckTimer = 0.0f;
+            UpdateCameraCache();
+            isCurrentlyVisible = IsVisible();
+
+            // Toggle water plane visibility
+            if (waterPlane.Visible != (isCurrentlyVisible && isEffectEnabled))
+            {
+                waterPlane.Visible = isCurrentlyVisible && isEffectEnabled;
+            }
```

Comment:
Isn't this kind of unnecessary? If I remember right IsVisible just returns the Godot property `Visible` which is basically a visibility flag on this node itself, which doesn't seem all that important to check? And also the water plane will be invisible anyway if this node (its parent) is invisible. Or did I misunderstand something here?